src/run_graph.rs

```diff
@@ -218,11 +257,25 @@ pub fn run_ex<DB: WriteResults + Sync>(
             let join = scope.builder().name(name).spawn(|| -> Result<()> {
                 // This uses a `loop` instead of a `while let` to avoid locking the graph too much
                 loop {
-                    let option_task = graph.lock().unwrap().next_task();
+                    let option_task = graph.lock().unwrap().next_task(ex, db);
                     if let Some((id, task)) = option_task {
                         info!("running task: {:?}", task);
-                        task.run(ex, db)?;
-                        graph.lock().unwrap().mark_as_completed(id);
+                        if let Err(e) = task.run(config, ex, db) {
+                            error!("task failed, marking childs as failed too: {:?}", task);
+                            util::report_error(&e);
+
+                            let result = if config.is_broken(&task.krate) {
+                                TestResult::BuildFail
+                            } else {
+                                TestResult::Error
+                            };
+                            graph
+                                .lock()
+                                .unwrap()
+                                .mark_as_failed(id, ex, db, &e, result)?;
+                        } else {
+                            graph.lock().unwrap().mark_as_completed(id);
+                        }
                     } else {
                         break;
```

Comment:
I'd add a comment here to explicitly state that "Threads terminate as soon as there's no available work, which is subideal in general but fine for crater today because the dependency chains are linear, so the completion of one task will never permit more than one more to be executed."

(assuming I'm remembering correctly)